tla/ccfraft.tla

```diff
@@ -121,98 +132,248 @@ VARIABLE configurations
 \* TODO: re-add a server in a raft_scenario test?
 VARIABLE removedFromConfiguration
 
-reconfigurationVars == <<reconfigurationCount, removedFromConfiguration, configurations>>
+RemovedFromConfigurationTypeInv ==
+    removedFromConfiguration \subseteq Servers
+
+reconfigurationVars == <<
+    reconfigurationCount, 
+    removedFromConfiguration, 
+    configurations
+>>
+
+ReconfigurationVarsTypeInv ==
+    /\ ReconfigurationCountTypeInv
+    /\ ConfigurationsTypeInv
+    /\ RemovedFromConfigurationTypeInv
 
 \* A set representing requests and responses sent from one server
 \* to another. With CCF, we have message integrity and can ensure unique messages.
 \* Messages only records messages that are currently in-flight, actions should
 \* removed messages once received.
 VARIABLE messages
 
+\* Helper function for checking the type safety of log entries
+EntryTypeOK(entry) ==
+    /\ entry.term \in Nat \ {0}
+    /\ \/ /\ entry.contentType = TypeEntry
+          /\ entry.value \in Nat \ {0}
+       \/ /\ entry.contentType = TypeSignature
+          /\ entry.value = Nil
+       \/ /\ entry.contentType = TypeReconfiguration
+          /\ entry.value \subseteq Servers
+
+AppendEntriesRequestTypeOK(m) ==
+    /\ m.mtype = AppendEntriesRequest
+    /\ m.mprevLogIndex \in Nat
+    /\ m.mprevLogTerm \in Nat
+    /\ m.mcommitIndex \in Nat
+    /\ \A k \in DOMAIN m.mentries: EntryTypeOK(m.mentries[k])
+
+AppendEntriesResponseTypeOK(m) ==
+    /\ m.mtype = AppendEntriesResponse
+    /\ m.msuccess \in BOOLEAN
+    /\ m.mmatchIndex \in Nat
```

Comment:
Should we try to line up with implementation names? This is called `last_log_idx` in the code.

Reply:
`last_log_idx` and `matchIndex` aren't quite the same so I wonder if its more confusing the use the same name of different things

Reply:
I am perhaps confused here, but assuming this describes a field in AppendEntriesResponse, it is called `last_log_idx` in the implementation. `match_idx` in the implementation is something the primary maintains per node, and you're right that it isn't always identical to the latest `last_log_idx` returned by a backup.

Are you saying that the spec AERs compute `matchIdx` and return that rather than `last_log_idx`? If so, that sounds like it might complicate trace validation.

Reply:
Related to `matchIndex` and `last_log_idx`: https://github.com/microsoft/CCF/issues/5057#issuecomment-1512420114

Reply:
Yes, it describes the field in `AppendEntriesResponse` (and yep, there is also a `matchIndex` in the node state). In the spec, the `matchIndex` in AERs is unused for NACKs. The ccf code seems to do something different for both AERs and for node state (where it can go backwards?). Also the term in AERs also seems to have a different meaning from the spec. For now, I'll rename as suggested and circle back to this later

Reply:
Done in 80438e7cac02d313c4798c6e9902f531b9b088e3

Reply:
In a similar vein, `mcommitIdx` , `mprevLogTerm`, and `mprevLogIndex` of `AppendEntriesRequest` could be renamed to `leaderCommitIndex`, `prevTerm`, and `prevIndex`.